docs/rfcs/20220829-distributed-dynamic-routing.md

```diff
@@ -0,0 +1,224 @@
+
+- The implementation is relatively complex, and the realtime load needs to be collected and reported by the computing node, with a certain extra cost.
+- CeresMeta maintains a Map<Table, Tablecost> for each computing node. When the load of a node reaches the threshold, rebalance is triggered, some tables are selected, migrated to the computing node with the lowest current load, and the routing relationship between shard and table is adjusted.
+```
+    ┌──────┐        ┌──────┐         ┌──────┐    ┌───────┐    ┌──────┐        ┌──────┐         ┌──────┐    
+ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ │       │ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ 
+ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ │ Table │ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ 
+ │┌────────┐  │  │            │   │┌────────┐  │ │ Load  │ │┌────────┐  │  │            │   │┌────────┐  │ 
+ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ │Balance│ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ 
+ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ │       │ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ 
+ ││ table0  │ │  │└─────────┘ │   ││ table2  │ │ └───────┘ ││ table0  │ │  │└─────────┘ │   ││ table1  │ │ 
+ ││ table1  │ │  │┌─────────┐ │   ││ table3  │ │──────────▶││         │ │  │┌─────────┐ │   ││ table2  │ │ 
+ ││         │ │  ││Shard1 F │ │   ││         │ │           ││         │ │  ││Shard1 F │ │   ││ table3  │ │ 
+ │└─────────┘ │  │└─────────┘ │   │└─────────┘ │           │└─────────┘ │  │└─────────┘ │   │└─────────┘ │ 
+ │┌─────────┐ │  │            │   │┌─────────┐ │           │┌─────────┐ │  │            │   │┌─────────┐ │ 
+ ││Shard2 L │ │  │            │   ││Shard2 F │ │           ││Shard2 L │ │  │            │   ││Shard2 F │ │ 
+ │└─────────┘ │  │            │   │└─────────┘ │           │└─────────┘ │  │            │   │└─────────┘ │ 
+┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐         ┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐
+│  CPUUse=90%  │                 │  CPUUse=30%  │         │  CPUUse=50%  │                 │  CPUUse=50%  │
+└──────────────┘                 └──────────────┘         └──────────────┘                 └──────────────┘
+```
+
+### Shard Change Process
+#### Switch Leader
+
+1. First, set the original leader shard and the follwer to be switched to PENDING_ FOLLOWER and PENDING_ LEADER.
+1. Wait for the leader shard to complete the preparatory action before the leader switch, and notify the CeresMeta after the completion. In this process, the new leader cannot process the write request.
+1. After receiving this request, meta will update the status of the two shards, and the leader switch is completed.
+
+
+#### Shard Migrate
+
+- Leader Shard：
```

Comment:
Maybe we have to take into considerations the no follower cluster mode (replication_factor == 1).

For the case where replication_factor == 1, all the shards should follow one procedure:
```
1. Close the shard on the old shard;
2. Open the shard on the new shard;
3. Adjust shard node mapping;
```

Reply:
I missed replication_factor, I always default that each leader shard will have multiple followers. I will distinguish and supplement the specific descriptions of these two situations in all the designs of shard.

Reply:
> Maybe we have to take into considerations the no follower cluster mode (replication_factor == 1).
> 
> For the case where replication_factor == 1, all the shards should follow one procedure:
> 
> ```
> 1. Close the shard on the old shard;
> 2. Open the shard on the new shard;
> 3. Adjust shard node mapping;
> ```

I missed replication_factor, I always default that each leader shard will have multiple followers. I will distinguish and supplement the specific descriptions of these two situations in all the designs of shard.